File: unitesting-framework-java/src/test/java/com/delta/TestUspDeltaFolders.java
package com.delta;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * TestUspDeltaFolders
 * ===================
 * Unit tests for the {@code [dbo].[USP_DELTA_FOLDERS]} stored procedure.
 *
 * Each test exercises one distinct behaviour path of the SP. The
 * {@code cleanTables()} method in {@link BaseTest} ensures a clean state
 * before every test.
 *
 * Test catalogue
 * --------------
 * TC01  Empty staging table             → early-exit with SUCCESS, input_row_count=0
 * TC02  Single new row                  → inserted into TARGET_FOLDERS & HASH_REGISTRY
 * TC03  Staging row marked processed    → Is_Processed flipped to 'Y'
 * TC04  Duplicate rows deduplication    → only first occurrence processed; rows_deduped counted
 * TC05  Unchanged row (hash pre-exists) → rows_unchanged=1, no new TARGET insert
 * TC06  Old hashes retired              → rows not in current batch deleted from HASH_REGISTRY
 * TC07  Metrics captured in log         → all counters correct for a mixed batch
 * TC08  Already-processed rows ignored  → Is_Processed='Y' rows not re-counted
 * TC09  Mixed batch (new + unchanged)   → correct per-category counts
 * TC10  Log entry entity name           → SYS_DELTA_LOG always records entity_name='FOLDERS'
 */
@DisplayName("USP_DELTA_FOLDERS Tests")
class TestUspDeltaFolders extends BaseTest {

    // -------------------------------------------------------------------------
    // TC01 – Empty staging table
    // -------------------------------------------------------------------------
    @Test
    @DisplayName("TC01: Empty staging table logs SUCCESS with zero counts")
    void testEmptyStagingLogsSuccess() throws Exception {
        DbUtils.execUspDeltaFolders(conn);

        Map<String, Object> log = DbUtils.getLatestLog(conn);
        assertNotNull(log, "Expected a log entry to be created");
        assertEquals("SUCCESS", log.get("status"));
        assertEquals(0, log.get("input_row_count"));
        assertEquals(0, log.get("rows_upserted"));
        assertEquals(0, log.get("rows_unchanged"));
        assertEquals(0, log.get("rows_deduped"));
        assertEquals(0, log.get("rows_retired"));
    }

    // -------------------------------------------------------------------------
    // TC02 – Single new row inserted into TARGET_FOLDERS
    // -------------------------------------------------------------------------
    @Test
    @DisplayName("TC02: Single new row is inserted into TARGET_FOLDERS and HASH_REGISTRY")
    void testSingleNewRowInsertedToTarget() throws Exception {
        long stagingId = DbUtils.insertStagingParent(conn);
        DbUtils.insertStagingFolder(conn, stagingId);

        DbUtils.execUspDeltaFolders(conn);

        assertEquals(1, DbUtils.getTargetFoldersCount(conn));
        assertEquals(1, DbUtils.getHashRegistryCount(conn));

        Map<String, Object> log = DbUtils.getLatestLog(conn);
        assertEquals("SUCCESS", log.get("status"));
        assertEquals(1, log.get("input_row_count"));
        assertEquals(1, log.get("rows_upserted"));
        assertEquals(0, log.get("rows_unchanged"));
    }

    // -------------------------------------------------------------------------
    // TC03 – Staging row Is_Processed flag is set to 'Y'
    // -------------------------------------------------------------------------
    @Test
    @DisplayName("TC03: All staging rows are marked Is_Processed='Y' after a successful run")
    void testStagingRowMarkedAsProcessed() throws Exception {
        long stagingId = DbUtils.insertStagingParent(conn);
        DbUtils.insertStagingFolder(conn, stagingId, "F001");
        DbUtils.insertStagingFolder(conn, stagingId, "F002");

        assertEquals(2, DbUtils.getUnprocessedStagingCount(conn));

        DbUtils.execUspDeltaFolders(conn);

        assertEquals(2, DbUtils.getProcessedStagingCount(conn));
        assertEquals(0, DbUtils.getUnprocessedStagingCount(conn));
    }

    // -------------------------------------------------------------------------
    // TC04 – Duplicate rows in staging are deduplicated
    // -------------------------------------------------------------------------
    @Test
    @DisplayName("TC04: Three identical staging rows → one TARGET insert, rows_deduped=2")
    void testDuplicateStagingRowsDeduped() throws Exception {
        long stagingId = DbUtils.insertStagingParent(conn);
        for (int i = 0; i < 3; i++) {
            DbUtils.insertStagingFolder(
                conn, stagingId,
                null, "DUP-001", "ACTIVE", "TYPE-A", "Duplicate", "ENTITY-001", "N");
        }

        DbUtils.execUspDeltaFolders(conn);

        assertEquals(1, DbUtils.getTargetFoldersCount(conn));
        assertEquals(1, DbUtils.getHashRegistryCount(conn));

        Map<String, Object> log = DbUtils.getLatestLog(conn);
        assertEquals(3, log.get("input_row_count"));
        assertEquals(2, log.get("rows_deduped"));
        assertEquals(1, log.get("rows_upserted"));
    }

    // -------------------------------------------------------------------------
    // TC05 – Unchanged row (hash already in HASH_REGISTRY)
    // -------------------------------------------------------------------------
    @Test
    @DisplayName("TC05: Row whose hash already exists → rows_unchanged=1, no new TARGET row")
    void testUnchangedRowNotReinserted() throws Exception {
        String folderFolderId = "KNOWN-001";
        long stagingId = DbUtils.insertStagingParent(conn);
        DbUtils.insertStagingFolder(conn, stagingId, folderFolderId);

        // Pre-insert the matching hash to simulate a previous run
        byte[] rowHash = DbUtils.computeFolderHash(
            conn, null, folderFolderId, "ACTIVE", "TYPE-A", "Test Folder", "ENTITY-001");
        DbUtils.insertHashRegistry(conn, "FOLDERS", rowHash, 60);

        DbUtils.execUspDeltaFolders(conn);

        assertEquals(0, DbUtils.getTargetFoldersCount(conn),
            "Hash already known – no new TARGET row");
        assertEquals(1, DbUtils.getHashRegistryCount(conn),
            "Hash registry count must remain 1");

        Map<String, Object> log = DbUtils.getLatestLog(conn);
        assertEquals(1, log.get("rows_unchanged"));
        assertEquals(0, log.get("rows_upserted"));
    }

    // -------------------------------------------------------------------------
    // TC06 – Old hashes are retired
    // -------------------------------------------------------------------------
    @Test
    @DisplayName("TC06: Hash from a prior run absent in this batch is deleted (retired)")
    void testOldHashesRetired() throws Exception {
        byte[] staleHash = new byte[32]; // all-zero 32-byte value
        DbUtils.insertHashRegistry(conn, "FOLDERS", staleHash, 120);

        long stagingId = DbUtils.insertStagingParent(conn);
        DbUtils.insertStagingFolder(conn, stagingId, "NEW-AFTER-RETIRE");

        DbUtils.execUspDeltaFolders(conn);

        // After the run there should be exactly 1 hash (the new one)
        assertEquals(1, DbUtils.getHashRegistryCount(conn));

        Map<String, Object> log = DbUtils.getLatestLog(conn);
        assertEquals(1, log.get("rows_retired"), "Stale hash must be counted as retired");
    }

    // -------------------------------------------------------------------------
    // TC07 – All metrics captured correctly in a multi-row batch
    // -------------------------------------------------------------------------
    @Test
    @DisplayName("TC07: input_row_count, rows_upserted, rows_unchanged, rows_deduped, rows_retired "
               + "are all captured correctly")
    void testLogCapturesAllMetrics() throws Exception {
        long stagingId = DbUtils.insertStagingParent(conn);

        // One row whose hash pre-exists (→ unchanged)
        String existingFolderId = "EXIST-001";
        byte[] existingHash = DbUtils.computeFolderHash(
            conn, null, existingFolderId, "ACTIVE", "TYPE-A", "Test Folder", "ENTITY-001");
        DbUtils.insertHashRegistry(conn, "FOLDERS", existingHash, 60);
        DbUtils.insertStagingFolder(conn, stagingId, existingFolderId);

        // Two new distinct rows (→ upserted=2)
        DbUtils.insertStagingFolder(conn, stagingId, "NEW-001");
        DbUtils.insertStagingFolder(conn, stagingId, "NEW-002");

        // One duplicate of NEW-001 (→ deduped=1)
        DbUtils.insertStagingFolder(conn, stagingId, "NEW-001");

        DbUtils.execUspDeltaFolders(conn);

        Map<String, Object> log = DbUtils.getLatestLog(conn);
        assertEquals("SUCCESS", log.get("status"));
        assertEquals(4, log.get("input_row_count"));  // 1 existing + 2 new + 1 dup
        assertEquals(1, log.get("rows_unchanged"));
        assertEquals(2, log.get("rows_upserted"));
        assertEquals(1, log.get("rows_deduped"));
        assertEquals(0, log.get("rows_retired"));      // existing hash was updated, not retired
    }

    // -------------------------------------------------------------------------
    // TC08 – Already-processed rows are ignored
    // -------------------------------------------------------------------------
    @Test
    @DisplayName("TC08: Rows with Is_Processed='Y' are not counted by the SP")
    void testAlreadyProcessedRowsIgnored() throws Exception {
        long stagingId = DbUtils.insertStagingParent(conn);
        DbUtils.insertStagingFolder(
            conn, stagingId, null, "OLD-Y", "ACTIVE", "TYPE-A", "Test Folder", "ENTITY-001", "Y");
        DbUtils.insertStagingFolder(
            conn, stagingId, null, "OLD-Y2", "ACTIVE", "TYPE-A", "Test Folder", "ENTITY-001", "Y");

        DbUtils.execUspDeltaFolders(conn);

        Map<String, Object> log = DbUtils.getLatestLog(conn);
        assertEquals("SUCCESS", log.get("status"));
        assertEquals(0, log.get("input_row_count"));   // SP sees nothing unprocessed
        assertEquals(0, DbUtils.getTargetFoldersCount(conn));
    }

    // -------------------------------------------------------------------------
    // TC09 – Mixed batch: new rows + unchanged rows
    // -------------------------------------------------------------------------
    @Test
    @DisplayName("TC09: 2 new rows + 1 pre-existing hash → upserted=2, unchanged=1")
    void testMixedNewAndUnchangedRows() throws Exception {
        long stagingId = DbUtils.insertStagingParent(conn);

        // Pre-existing hash
        String knownId = "MIX-KNOWN";
        byte[] knownHash = DbUtils.computeFolderHash(
            conn, null, knownId, "ACTIVE", "TYPE-A", "Test Folder", "ENTITY-001");
        DbUtils.insertHashRegistry(conn, "FOLDERS", knownHash, 60);
        DbUtils.insertStagingFolder(conn, stagingId, knownId);

        // Two genuinely new rows
        DbUtils.insertStagingFolder(conn, stagingId, "MIX-NEW-1");
        DbUtils.insertStagingFolder(conn, stagingId, "MIX-NEW-2");

        DbUtils.execUspDeltaFolders(conn);

        assertEquals(2, DbUtils.getTargetFoldersCount(conn));

        Map<String, Object> log = DbUtils.getLatestLog(conn);
        assertEquals(2, log.get("rows_upserted"));
        assertEquals(1, log.get("rows_unchanged"));
        assertEquals(3, log.get("input_row_count"));
    }

    // -------------------------------------------------------------------------
    // TC10 – Log entry always records the correct entity name
    // -------------------------------------------------------------------------
    @Test
    @DisplayName("TC10: SYS_DELTA_LOG always records entity_name='FOLDERS'")
    void testLogEntryRecordsEntityName() throws Exception {
        DbUtils.execUspDeltaFolders(conn);

        Map<String, Object> log = DbUtils.getLatestLog(conn);
        assertNotNull(log);
        assertEquals("FOLDERS", log.get("entity_name"));
        assertTrue(DbUtils.getLogCount(conn) >= 1);
    }
}

File: unitesting-framework-main/tests/test_usp_delta_folders.py
"""
test_usp_delta_folders.py
=========================
Unit tests for the [dbo].[USP_DELTA_FOLDERS] stored procedure.

Each test exercises one distinct behaviour path of the SP.  The
``clean_tables`` autouse fixture (conftest.py) ensures a clean state
before every test.

Test catalogue
--------------
TC01  Empty staging table            → early-exit with SUCCESS, input_row_count=0
TC02  Single new row                 → inserted into TARGET_FOLDERS & HASH_REGISTRY
TC03  Staging row marked processed   → Is_Processed flipped to 'Y'
TC04  Duplicate rows deduplication   → only first occurrence processed; rows_deduped counted
TC05  Unchanged row (hash pre-exists)→ rows_unchanged=1, no new TARGET insert
TC06  Old hashes retired             → rows not in current batch deleted from HASH_REGISTRY
TC07  Metrics captured in log        → all counters correct for a mixed batch
TC08  Already-processed rows ignored → Is_Processed='Y' rows not re-counted
TC09  Mixed batch (new + unchanged)  → correct per-category counts
TC10  Log entry entity name          → SYS_DELTA_LOG always records entity_name='FOLDERS'
"""

from __future__ import annotations

import pytest

from tests.helpers.db_utils import (
    compute_folder_hash,
    exec_usp_delta_folders,
    get_hash_registry_count,
    get_latest_log,
    get_log_count,
    get_processed_staging_count,
    get_target_folders_count,
    get_unprocessed_staging_count,
    insert_hash_registry,
    insert_staging_folder,
    insert_staging_parent,
)


class TestUSPDeltaFolders:
    """Integration tests for USP_DELTA_FOLDERS."""

    # ------------------------------------------------------------------
    # TC01 – Empty staging table
    # ------------------------------------------------------------------
    def test_empty_staging_logs_success(self, db_conn):
        """TC01: With no unprocessed rows the SP exits early with SUCCESS."""
        cursor = db_conn.cursor()

        exec_usp_delta_folders(cursor)

        log = get_latest_log(cursor)
        assert log is not None, "Expected a log entry to be created"
        assert log["status"] == "SUCCESS"
        assert log["input_row_count"] == 0
        assert log["rows_upserted"] == 0
        assert log["rows_unchanged"] == 0
        assert log["rows_deduped"] == 0
        assert log["rows_retired"] == 0

        cursor.close()

    # ------------------------------------------------------------------
    # TC02 – Single new row inserted into TARGET_FOLDERS
    # ------------------------------------------------------------------
    def test_single_new_row_inserted_to_target(self, db_conn):
        """TC02: One new staging row → inserted into TARGET_FOLDERS and HASH_REGISTRY."""
        cursor = db_conn.cursor()

        staging_id = insert_staging_parent(cursor)
        insert_staging_folder(cursor, staging_id=staging_id)

        exec_usp_delta_folders(cursor)

        assert get_target_folders_count(cursor) == 1
        assert get_hash_registry_count(cursor) == 1

        log = get_latest_log(cursor)
        assert log["status"] == "SUCCESS"
        assert log["input_row_count"] == 1
        assert log["rows_upserted"] == 1
        assert log["rows_unchanged"] == 0

        cursor.close()

    # ------------------------------------------------------------------
    # TC03 – Staging row Is_Processed flag is set to 'Y'
    # ------------------------------------------------------------------
    def test_staging_row_marked_as_processed(self, db_conn):
        """TC03: After a successful run all staging rows are flipped to Is_Processed='Y'."""
        cursor = db_conn.cursor()

        staging_id = insert_staging_parent(cursor)
        insert_staging_folder(cursor, staging_id=staging_id, folder_folder_id="F001")
        insert_staging_folder(cursor, staging_id=staging_id, folder_folder_id="F002")

        assert get_unprocessed_staging_count(cursor) == 2

        exec_usp_delta_folders(cursor)

        assert get_processed_staging_count(cursor) == 2
        assert get_unprocessed_staging_count(cursor) == 0

        cursor.close()

    # ------------------------------------------------------------------
    # TC04 – Duplicate rows in staging are deduplicated
    # ------------------------------------------------------------------
    def test_duplicate_staging_rows_deduped(self, db_conn):
        """TC04: Three identical rows → one TARGET insert; rows_deduped=2."""
        cursor = db_conn.cursor()

        staging_id = insert_staging_parent(cursor)
        for _ in range(3):
            insert_staging_folder(
                cursor,
                staging_id=staging_id,
                folder_folder_id="DUP-001",
                folder_description="Duplicate",
            )

        exec_usp_delta_folders(cursor)

        assert get_target_folders_count(cursor) == 1
        assert get_hash_registry_count(cursor) == 1

        log = get_latest_log(cursor)
        assert log["input_row_count"] == 3
        assert log["rows_deduped"] == 2
        assert log["rows_upserted"] == 1

        cursor.close()

    # ------------------------------------------------------------------
    # TC05 – Unchanged row (hash already in HASH_REGISTRY)
    # ------------------------------------------------------------------
    def test_unchanged_row_not_reinserted(self, db_conn):
        """TC05: A row whose hash already exists → rows_unchanged=1, no new TARGET row."""
        cursor = db_conn.cursor()

        folder_folder_id = "KNOWN-001"
        staging_id = insert_staging_parent(cursor)
        insert_staging_folder(
            cursor,
            staging_id=staging_id,
            folder_folder_id=folder_folder_id,
        )

        # Pre-insert the matching hash to simulate a previous run
        row_hash = compute_folder_hash(
            cursor,
            None,            # folderIfUnmodifiedSince
            folder_folder_id,
            "ACTIVE",
            "TYPE-A",
            "Test Folder",
            "ENTITY-001",
        )
        insert_hash_registry(cursor, "FOLDERS", row_hash, age_seconds=60)

        exec_usp_delta_folders(cursor)

        assert get_target_folders_count(cursor) == 0, "Hash already known – no new TARGET row"
        assert get_hash_registry_count(cursor) == 1, "Hash registry count must remain 1"

        log = get_latest_log(cursor)
        assert log["rows_unchanged"] == 1
        assert log["rows_upserted"] == 0

        cursor.close()

    # ------------------------------------------------------------------
    # TC06 – Old hashes are retired
    # ------------------------------------------------------------------
    def test_old_hashes_retired(self, db_conn):
        """TC06: A hash from a prior run that is absent in this batch is deleted."""
        cursor = db_conn.cursor()

        # Insert a stale hash into the registry (not part of the current batch)
        stale_hash = b"\x00" * 32  # Arbitrary 32-byte value
        insert_hash_registry(cursor, "FOLDERS", stale_hash, age_seconds=120)

        # New, different staging row
        staging_id = insert_staging_parent(cursor)
        insert_staging_folder(
            cursor,
            staging_id=staging_id,
            folder_folder_id="NEW-AFTER-RETIRE",
        )

        exec_usp_delta_folders(cursor)

        # After the run there should be exactly 1 hash (the new one)
        assert get_hash_registry_count(cursor) == 1

        log = get_latest_log(cursor)
        assert log["rows_retired"] == 1, "Stale hash must be counted as retired"

        cursor.close()

    # ------------------------------------------------------------------
    # TC07 – All metrics captured correctly in a multi-row batch
    # ------------------------------------------------------------------
    def test_log_captures_all_metrics(self, db_conn):
        """TC07: input_row_count, rows_upserted, rows_unchanged, rows_deduped, rows_retired."""
        cursor = db_conn.cursor()

        staging_id = insert_staging_parent(cursor)

        # One row whose hash pre-exists (→ unchanged)
        existing_folder_id = "EXIST-001"
        existing_hash = compute_folder_hash(
            cursor, None, existing_folder_id, "ACTIVE", "TYPE-A", "Test Folder", "ENTITY-001"
        )
        insert_hash_registry(cursor, "FOLDERS", existing_hash, age_seconds=60)
        insert_staging_folder(cursor, staging_id=staging_id, folder_folder_id=existing_folder_id)

        # Two new distinct rows (→ upserted=2)
        insert_staging_folder(cursor, staging_id=staging_id, folder_folder_id="NEW-001")
        insert_staging_folder(cursor, staging_id=staging_id, folder_folder_id="NEW-002")

        # One duplicate of NEW-001 (→ deduped=1)
        insert_staging_folder(cursor, staging_id=staging_id, folder_folder_id="NEW-001")

        exec_usp_delta_folders(cursor)

        log = get_latest_log(cursor)
        assert log["status"] == "SUCCESS"
        assert log["input_row_count"] == 4   # 1 existing + 2 new + 1 dup
        assert log["rows_unchanged"] == 1
        assert log["rows_upserted"] == 2
        assert log["rows_deduped"] == 1
        assert log["rows_retired"] == 0      # existing hash was updated, not retired

        cursor.close()

    # ------------------------------------------------------------------
    # TC08 – Already-processed rows are ignored
    # ------------------------------------------------------------------
    def test_already_processed_rows_ignored(self, db_conn):
        """TC08: Rows with Is_Processed='Y' are not counted by the SP."""
        cursor = db_conn.cursor()

        staging_id = insert_staging_parent(cursor)
        # Insert rows that are already processed
        insert_staging_folder(
            cursor, staging_id=staging_id, folder_folder_id="OLD-Y", is_processed="Y"
        )
        insert_staging_folder(
            cursor, staging_id=staging_id, folder_folder_id="OLD-Y2", is_processed="Y"
        )

        exec_usp_delta_folders(cursor)

        log = get_latest_log(cursor)
        assert log["status"] == "SUCCESS"
        assert log["input_row_count"] == 0    # SP sees nothing unprocessed
        assert get_target_folders_count(cursor) == 0

        cursor.close()

    # ------------------------------------------------------------------
    # TC09 – Mixed batch: new rows + unchanged rows
    # ------------------------------------------------------------------
    def test_mixed_new_and_unchanged_rows(self, db_conn):
        """TC09: 2 new rows + 1 pre-existing hash → upserted=2, unchanged=1."""
        cursor = db_conn.cursor()

        staging_id = insert_staging_parent(cursor)

        # Pre-existing hash
        known_id = "MIX-KNOWN"
        known_hash = compute_folder_hash(
            cursor, None, known_id, "ACTIVE", "TYPE-A", "Test Folder", "ENTITY-001"
        )
        insert_hash_registry(cursor, "FOLDERS", known_hash, age_seconds=60)
        insert_staging_folder(cursor, staging_id=staging_id, folder_folder_id=known_id)

        # Two genuinely new rows
        insert_staging_folder(cursor, staging_id=staging_id, folder_folder_id="MIX-NEW-1")
        insert_staging_folder(cursor, staging_id=staging_id, folder_folder_id="MIX-NEW-2")

        exec_usp_delta_folders(cursor)

        assert get_target_folders_count(cursor) == 2

        log = get_latest_log(cursor)
        assert log["rows_upserted"] == 2
        assert log["rows_unchanged"] == 1
        assert log["input_row_count"] == 3

        cursor.close()

    # ------------------------------------------------------------------
    # TC10 – Log entry always records the correct entity name
    # ------------------------------------------------------------------
    def test_log_entry_records_entity_name(self, db_conn):
        """TC10: SYS_DELTA_LOG rows always carry entity_name='FOLDERS'."""
        cursor = db_conn.cursor()

        exec_usp_delta_folders(cursor)

        log = get_latest_log(cursor)
        assert log is not None
        assert log["entity_name"] == "FOLDERS"
        assert get_log_count(cursor) >= 1

        cursor.close()
